packages/tools/scripts/rewrite-internal-redirects/index.ts

```diff
@@ -0,0 +1,307 @@
+          const destCompile = pathToRegexp.compile(destUrl.pathname)
+
+          const newPath = destCompile(params)
+
+          destUrl.pathname = newPath
+
+          return destUrl.href
+        }
+      }
+    } else {
+      destination = pathToRegexp.compile(redirect.destination)
+    }
+
+    return {
+      source: pathToRegexp.match(redirect.source),
+      destination,
+      permanent: redirect.permanent,
+    }
+  })
+
+  return redirects
+}
+
+/**
+ * Checks for a matching redirect with the given URL and, if found,
+ * applies the matching redirect.
+ *
+ * @param url URL to check for redirects with and apply to
+ * @param redirects List of redirects which will be tested against the URL
+ * @returns The redirected URL
+ */
+export const checkAndApplyRedirect = (
+  url: string,
+  redirects
+): string | false => {
+  let matchedResult: pathToRegexp.Match = false
+  let matchedRedirect
+
+  const [urlWithoutHash, hash] = url.split('#')
+
+  redirects.some((redirect) => {
+    if (!redirect.permanent) return
+    matchedResult = redirect.source(url) || redirect.source(urlWithoutHash)
+    if (matchedResult) {
+      matchedRedirect = redirect
+      return true
+    }
+  })
+
+  if (matchedRedirect && matchedResult) {
+    // If the matched destination has no tokens, we can just return it
+    if (typeof matchedRedirect.destination === 'string') {
+      // include new hash
+      if (matchedRedirect.destination.includes('#')) {
+        return matchedRedirect.destination
+      }
+      // include old hash, if exists
+      return `${matchedRedirect.destination}${hash ? `#${hash}` : ''}`
+    }
+
+    // TS is not cooperating, so having to use typecasting here
+    const destinationUrl = matchedRedirect.destination(
+      typeof (matchedResult as pathToRegexp.Match) !== 'boolean'
+        ? (matchedResult as unknown as pathToRegexp.MatchResult).params
+        : {}
+    )
+
+    return `${destinationUrl}${hash ? `#${hash}` : ''}`
+  }
+
+  return false
+}
+
+/**
+ * Remark plugin which accepts a list of redirects and applies them to any matching links
+ */
+const rewriteInternalRedirectsPlugin = ({ product, redirects }) => {
+  return async function transformer(tree, file) {
+    return visit(tree, (node: Node & { url?: string }) => {
+      if (!is(node, 'link') && !is(node, 'definition')) return
+
+      // Only check internal links
+      if (node.url && !node.url.startsWith('#') && isInternalUrl(node.url)) {
+        const urlToRedirect = node.url.startsWith('/')
+          ? node.url
+          : new URL(node.url).pathname
+
+        let redirectUrl
+
+        if (isInternalUrl(node.url, product)) {
+          redirectUrl = checkAndApplyRedirect(urlToRedirect, redirects)
+        } else {
+          const [, hash] = node.url.split('#')
+          redirectUrl = fetch(node.url, { method: 'HEAD' }).then(
+            (res) =>
+              `${res.url}${hash && !res.url.includes('#') ? `#${hash}` : ''}`
+          )
+        }
```

Comment:
@thiskevinwang here's the core internal link detection / redirect check flow

Reply:
For terraform, we will want to adjust the logic for internal urls to prepend the domain and execute the fetch, then resolve.